atlasdb-impl-shared/src/main/java/com/palantir/atlasdb/transaction/impl/expectations/KeyValueServiceDataTracker.java

```diff
@@ -0,0 +1,129 @@
+     * This is un-synchronized as it is called after task completion and the abort/commit stage.
+     * For returned statistics, no consistency guarantees are provided if the user interacts with the transaction
+     * outside the context of {@link com.palantir.atlasdb.transaction.api.TransactionTask} or after commit/abort
+     * stages (e.g. this can happen with futures/iterators).
+     */
+    public ImmutableMap<TableReference, TransactionReadInfo> getReadInfoByTable() {
+        Set<TableReference> tableRefs = ImmutableSet.copyOf(Sets.intersection(
+                bytesReadByTable.asMap().keySet(), kvsCallByTable.asMap().keySet()));
+
+        return tableRefs.stream()
+                .collect(ImmutableMap.toImmutableMap(
+                        Function.identity(), tableRef -> ImmutableTransactionReadInfo.builder()
+                                .bytesRead(bytesReadByTable.get(tableRef))
+                                .kvsCalls(kvsCallByTable.get(tableRef))
+                                .maximumBytesKvsCallInfo(
+                                        Optional.ofNullable(maximumBytesKvsCallInfoByTable.get(tableRef)))
+                                .build()));
+    }
+
+    /**
+     * Tracks an effectively completed kvs read method call for some {@link TableReference}.
+     * Effectively completed refers to an eager call (i.e. does not spawn futures or iterators for later consumption).
+     */
+    public void readForTable(TableReference tableRef, String methodName, long bytesRead) {
+        KvsCallReadInfo callInfo = ImmutableKvsCallReadInfo.builder()
+                .bytesRead(bytesRead)
+                .methodName(methodName)
+                .build();
+        updateKvsMethodOverallTallies(callInfo);
+        updateKvsMethodByTableTallies(tableRef, callInfo);
+    }
+
+    /**
+     * Track a lazy kvs read method call for some {@link TableReference}.
+     */
+    public void partialReadForTable(TableReference tableRef, long bytesRead) {
+        bytesReadOverall.add(bytesRead);
+        bytesReadByTable.addAndGet(tableRef, bytesRead);
+    }
+
+    /**
+     * Track that a kvs read method was called for some {@link TableReference}.
+     */
+    public void callForTable(TableReference tableRef) {
+        kvsCallsOverall.increment();
+        kvsCallByTable.incrementAndGet(tableRef);
+    }
+
+    /**
+     * Track an effectively completed kvs read method call with no {@link TableReference} information.
+     * Effectively completed refers to an eager call (i.e. does not spawn futures or iterators for later consumption).
+     */
+    public void tableAgnosticRead(String methodName, long bytesRead) {
+        KvsCallReadInfo callInfo = ImmutableKvsCallReadInfo.builder()
+                .bytesRead(bytesRead)
+                .methodName(methodName)
+                .build();
+        updateKvsMethodOverallTallies(callInfo);
+    }
+
+    private void updateKvsMethodOverallTallies(KvsCallReadInfo callInfo) {
```

Comment:
If you need to build kvs call read info everywhere just to pass it here, consider just having this method build the call info (or having callers do it)